design/pool-configuration.md

```diff
@@ -0,0 +1,359 @@
+# Address Pool configuration options
+
+## Summary
+
+Following @johananl's initiative in [this issue](https://github.com/metallb/metallb/issues/630)
+this enhancement tries to collect and classify various configuration enhancement requests
+requested over time.
+
+## Motivation
+
+The way advanced users consume MetalLB is evolving and exposed a need for more complex configuration.
+For example, the need for isolation and multi-tenancy induced requests for being able to advertise some specific address pools from a subset of the nodes, or
+to allocate them only to a specific namespace, to mention some.
+
+### Goals
+
+* Collect and classify all the requests coming from different issues, in order to have a better understanding
+  and an organic implementation, avoiding focus on and implementation of only a specific feature, risking causing conflicts afterwards.
+* Provide a list of the possible configuration enhancements and highlight dependencies and conflicts among them (if any).
+* Highlight the cases where the solution must be BGP or L2 specific
+
+### Non-Goals
+
+* Covering only a specific use case.
+* Providing a L2 only or BGP only solution where the problem might be expressed in a more generic way.
+
+## Proposal
+
+For all the IPs belonging to an address pool, we need to address three main configuration degrees:
```

Comment:
I am not sure an IP is the "central" entity we should think about for all cases.

For example, in most commercial routers, controlling which IP prefixes are advertised to which BGP peers is done at the **peer** level, not the IP level. You typically attach a "route map" to a BGP peer. The route map is basically a policy document which defines which prefixes known to the local BGP process are allowed to be advertised to the peer to which the route map is attached.

AFAICT we have the following concerns in MetalLB:

1. Which speaker nodes should announce a prefix. This would be the equivalent of selecting which routers to apply a configuration to.
2. Which BGP peers to advertise a prefix to. This would be the equivalent of restricting BGP advertisements using a route map.
3. Which IP to allocate to which service. This would be the equivalent of an IPAM policy that's **completely unrelated** to advertisement.

I'd consider rephrasing as follows:

>MetalLB deals with the following concerns:

Then I would rephrase the list to be more generic as follows:

- Controlling which speaker nodes advertise a prefix.
- Controlling which peers a speaker node should advertise a prefix to.
- Controlling which IPs are allocated to which services.

Lastly, see my [note](https://github.com/metallb/metallb/pull/942/files#r773898754) below.

Reply:
Sounds clearer, will apply your suggestion. I'll reply on the bgp side on the other comment below.